plugin/firewall/README.md

```diff
@@ -0,0 +1,112 @@
+# firewall
+
+## Name
+
+*firewall* - enables filtering on query and response using direct expression as policy.
+
+## Description
+
+The firewall plugin defines a rule list of expressions that triggers workflow action on the DNS query or its response.
+A rule list is an ordered set of rules that are evaluated in sequence.
+Each rule has two parts: an action and an expression. When the rule is evaluated,
+first the expression is evaluated.
+- If the expression evaluates to `true` the action is performed on the query and the rule list evaluation ceases.
+- If the expression does not evaluates to `true` then next rule in sequence is evaluated.
+
+
+## Syntax
+
+~~~ txt
+firewall DIRECTION {
+    ACTION EXPRESSION
+    ACTION EXPRESSION
+    ...
+}
+~~~~
+
+* **DIRECTION** indicates if the _rule list_ will be applied to queries or responses. It can be `query` or `response`.
+
+* **ACTION** defines the workflow action to apply to the DNS operation if the **EXPRESSION** evaluates to the boolean `true`
+action is one of:
+  - `allow` : continue the DNS resolution process
+  - `refuse` : interrupt the DNS resolution, reply with REFUSE code
+  - `block` : interrupt the DNS resolution, reply with NXDOMAIN code
+  - `drop` : interrupt the DNS resolution, do not reply
+
+* **EXPRESSION** defines the expression to evaluate in order to validate the action and interrupt the sequence of rules.
+Expression is a [go-like language](https://github.com/Knetic/govaluate/blob/master/MANUAL.md)
+where the variables are either the `metadata` of CoreDNS either a list of names associated with the DNS query/response information.
+Usual operators applies.
+
+Exemples of expression using the usual metadata:
+* `client_ip == '10.0.0.20'`
+* `type == 'AAAA'`
+* `type IN ('AAAA', 'A', 'TXT')`
+* `type IN ('AAAA', 'A') && name =~ 'google.com'`
+* `[mac/address] =~ '.*:FF:.*'`
+
+NOTE: because of the `/` separator included in a label of metadata, those labels must be enclosed on bracket [...] for a correct evaluation by the expression engine
+
+The following names are supported for querying information
```

Comment:
Would link to the log's plugin and say we support the same syntax (plus a few new ones? Are there new ones?)

Reply:
The `>` idea was that those are header related.

Reply:
It is almost the same:
- but {combined} and {common} are not supported as it makes no sense
- we have few others: server_ip, server_port and response_ip
- we need for Firewall that those are not enclosed in braces {}.

the rqdata tool used here is expected to be reuse for `log` and then for `rewrite`, so we have a common reference to all names of data extracted from the query.

But, maybe we can wait it is really shared before having a common reference in README ?